### PR TITLE
Use `max_upload_size` as the limit when following the `Location` header

### PR DESCRIPTION
**Pull request by S7evinK**
Otherwise we use the `expected_size` from the initial federation request, which might be far too low.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by S7evinK** [comment]
Complement failing seems unrelated.

**Review by erikjohnston — APPROVED** [review]

**Comment by erikjohnston** [comment]
Though a regression test would be good

File: tests/http/test_matrixfederationclient.py [hunk 1]
```diff
@@ -337,6 +340,71 @@ def test_client_gets_headers(self) -> None:
...
+
+        redirect_data = b"\r\n\r\n--6067d4698f8d40a0a794ea7d7379d53a\r\nContent-Type: application/json\r\n\r\n{}\r\n--6067d4698f8d40a0a794ea7d7379d53a\r\nLocation: http://testserv:8008/ab/c1/2345.txt\r\n\r\n--6067d4698f8d40a0a794ea7d7379d53a--\r\n\r\n"
+        client.dataReceived(
+            b"HTTP/1.1 200 OK\r\n"
+            b"Server: Fake\r\n"
+            b"Content-Length: %i\r\n"
+            b"Content-Type: multipart/mixed; boundary=6067d4698f8d40a0a794ea7d7379d53a\r\n\r\n"
+            % (len(redirect_data))
+        )
+        client.dataReceived(redirect_data)
```
Comment (S7evinK) [inline tests/http/test_matrixfederationclient.py:381]: I have no idea why this works, but it works, though.

**Review by erikjohnston — APPROVED** [review]
Otherwise lgtm

**Pull request merged**